fix: optimize read-inline-info

not-matthias opened the pull request.

not-matthias committed:
feat: optimize inline table lookups using hashtable and binary search

GuillaumeLagrange commented on .github/workflows/ci.yml:
```diff
@@ -22,6 +22,17 @@ jobs:
     steps:
       - uses: actions/checkout@v4
 
+      # Skip installing package docs to avoid wasting time when installing valgrind
```
VERY nitpicky: not "when installing valgrind", but rather "installing build dependencies"

GuillaumeLagrange commented on .github/workflows/codspeed.yml:
```diff
@@ -0,0 +1,84 @@
+          EOF
+
+      # Build and install Valgrind
+      #
```
Nitpick: unnecessary extra line

GuillaumeLagrange commented on coregrind/m_debuginfo/inltab_lookup.c:
```diff
@@ -0,0 +1,131 @@
+
+/* Public API for inline lookup table */
+
+void* VG_(inltab_lookup_new)(void) {
```
why return a void* here ? Is it to avoid exporting the lookup table type ?

Would make the code a tad bit safer and cleaner to understand to use `VgHashTable *` here and in other API functions

GuillaumeLagrange commented on coregrind/m_debuginfo/priv_inltab_lookup.h:
```diff
@@ -0,0 +1,12 @@
+extern void* VG_(inltab_lookup_new)(void);
+extern void VG_(inltab_lookup_insert)(void *lookup_table, Addr addr, Word inl_idx);
+extern Bool VG_(inltab_lookup_get)(void *lookup_table, Addr addr, Word *inl_idx,
+                                   const struct _DebugInfo* di);
```
Why is this `struct _DebufInfo*` but `DebugInfo*` in implementation?

I'm guessing it's typedefed somewhere to `DebugInfo`

not-matthias replied:
yep, it's typedefed in `pub_tool_debuginfo.h`. it's just a forward declaration so that we don't have to include it in this header as well. But it looks like we're already including it transitively in one of the headers so I changed it.

GuillaumeLagrange commented on .github/workflows/codspeed.yml:
```diff
@@ -17,7 +17,7 @@ jobs:
         valgrind: 
           - /usr/bin/valgrind
-          - ${{ github.workspace }}/vg-in-place
+          - ../vg-in-place
```
fixup with prev commit

GuillaumeLagrange reviewed (changes requested):
olgtm, will approve once we have at lesat tested teh impact compared to baseline valgrind.

Comments are pretty minor

art049 commented on Justfile:
```diff
@@ -0,0 +1,50 @@
+        -C /tmp/valgrind-upstream
+
+    # Build and install using build-in
+    just build-in "/tmp/valgrind-upstream/valgrind-{{ version }}"
```
What's the benefit of introducing a new build system while there is already automake?

not-matthias replied:
I added the Justfile so that we can have a matrix in CI which just calls `just build <version>` (with version being `local`, `3.26.0`, ...). This makes it la bit easier to understand rather than having conditional steps in the CI. Plus, we can also easily run it locally.

art049 replied:
ok, this could be done with make as well. not a big problem anyway.

not-matthias replied:
Yeah, I just prefer not putting too much logic into Github Actions to allow local execution and to reduce complexity. I'll keep it with the Justfile if you don't mind.

art049 commented on bench/bench.py:
```diff
@@ -0,0 +1,202 @@
+#!/usr/bin/env python3
```
We can make this an executable directly: `#!/usr/bin/env -S uv run --script`

https://docs.astral.sh/uv/guides/scripts/#using-a-shebang-to-create-an-executable-file

art049 commented on bench/bench.py:
I think this could be simpler with the following structure:

- just 1 test but taking various parameters:
  - the valgrind flags, you can use parameterize + ids to have it cleanly displayed in the bench uri, [see this](https://docs.pytest.org/en/stable/example/parametrize.html#set-marks-or-test-id-for-individual-parametrized-test)
  - the binary to instrument

then maybe we also need the valgrind version in the bench names?

not-matthias replied:
Ahhh, that's a good point. We could have the following uri: `test_{binary}_{valgrind_version}_{param_id}`

not-matthias replied:
We have this URI for the tests now: `test_valgrind[<valgrind_version>, <cmd>, <args_id>]` (e.g. `test_valgrind[valgrind-3.26.0, eza -alh, no-inline]`)

art049 reviewed (approved):
Code changes lgtm! Got some feedback on the devX though

art049 commented:
We also need a few addtional binaries to bench IMO

not-matthias commented:
> We also need a few addtional binaries to bench IMO

I've added ls and echo to the test binaries, lmk if you want anyting else as well. I've also modified the bench script a bit to allow passing an arbitrary cmd which will be useful for cmds with args. 

EDIT: Also added caching and disabled the take-strings benchmark for 3.25 since it takes a super long time.

GuillaumeLagrange reviewed (approved):
lgtm, very good work 🤗
